File: client/bigi/bigi_ffi.erl
-module(bigi_ffi).

-export([
    from/1,
    from_string/1,
    from_bytes/3,
    to/1,
    to_bytes/4,
    zero/0,
    one/0,
    n_one/0,
    ten/0,
    compare/2,
    negate/1,
    add/2,
    subtract/2,
    multiply/2,
    divide/2,
    divide_no_zero/2,
    remainder/2,
    remainder_no_zero/2,
    modulo/2,
    modulo_no_zero/2,
    power/2,
    decode/1,
    bitwise_and/2,
    bitwise_exclusive_or/2,
    bitwise_not/1,
    bitwise_or/2,
    bitwise_shift_left/2,
    bitwise_shift_right/2,
    from_base2/1,
    from_base8/1,
    from_base16/1,
    to_base2/1,
    to_base8/1,
    to_base16/1
]).

from(Int) -> Int.

from_string(Str) ->
    case string:to_integer(Str) of
        {_, Rest} when Rest /= <<"">> -> {error, nil};
        {Int, _} -> {ok, Int}
    end.

from_bytes(Bytes, Endianness, Signedness) ->
    BitSize = erlang:bit_size(Bytes),

    case BitSize rem 8 of
        0 ->
            case Endianness of
                little_endian ->
                    case Signedness of
                        signed ->
                            <<Int:BitSize/little-signed-integer>> = Bytes,
                            {ok, Int};
                        unsigned ->
                            <<Int:BitSize/little-unsigned-integer>> = Bytes,
                            {ok, Int}
                    end;
                big_endian ->
                    case Signedness of
                        signed ->
                            <<Int:BitSize/big-signed-integer>> = Bytes,
                            {ok, Int};
                        unsigned ->
                            <<Int:BitSize/big-unsigned-integer>> = Bytes,
                            {ok, Int}
                    end
            end;
        _ ->
            {error, nil}
    end.

to(BigInt) -> {ok, BigInt}.

to_bytes(BigInt, Endianness, Signedness, ByteCount) ->
    case ByteCount * 8 of
        BitCount when BitCount >= 8 ->
            RangeMin =
                case Signedness of
                    signed -> -(1 bsl (BitCount - 1));
                    unsigned -> 0
                end,

            RangeMax =
                case Signedness of
                    signed -> (1 bsl (BitCount - 1)) - 1;
                    unsigned -> (1 bsl BitCount) - 1
                end,

            % Error if the value is out of range for the available bits
            case BigInt >= RangeMin andalso BigInt =< RangeMax of
                true ->
                    case Endianness of
                        little_endian -> {ok, <<BigInt:BitCount/little-integer>>};
                        big_endian -> {ok, <<BigInt:BitCount/big-integer>>}
                    end;
                false ->
                    {error, nil}
            end;
        _ ->
            {error, nil}
    end.

zero() -> 0.
one() -> 1.
n_one() -> -1.
ten() -> 10.

compare(A, B) when A < B -> lt;
compare(A, B) when A > B -> gt;
compare(_, _) -> eq.

negate(A) -> -A.

add(A, B) -> A + B.

subtract(A, B) -> A - B.

multiply(A, B) -> A * B.

divide(_, 0) -> 0;
divide(A, B) -> A div B.

divide_no_zero(_, 0) -> {error, nil};
divide_no_zero(A, B) -> {ok, divide(A, B)}.

remainder(_, 0) -> 0;
remainder(A, B) -> A rem B.

remainder_no_zero(_, 0) -> {error, nil};
remainder_no_zero(A, B) -> {ok, remainder(A, B)}.

modulo(_, 0) -> 0;
modulo(A, B) -> ((A rem B) + B) rem B.

modulo_no_zero(_, 0) -> {error, nil};
modulo_no_zero(A, B) -> {ok, modulo(A, B)}.

power(_, Exp) when Exp < 0 -> {error, nil};
power(Base, Exp) -> {ok, do_power(Base, Exp)}.

do_power(_, 0) ->
    1;
do_power(A, 1) ->
    A;
do_power(A, N) ->
    B = do_power(A, N div 2),
    B * B *
        (case N rem 2 of
            0 -> 1;
            1 -> A
        end).

decode(Dyn) when is_integer(Dyn) -> {ok, Dyn};
decode(_Dyn) -> {error, 0}.

bitwise_and(A, B) -> A band B.

bitwise_exclusive_or(A, B) -> A bxor B.

bitwise_not(A) -> bnot A.

bitwise_or(A, B) -> A bor B.

bitwise_shift_left(A, B) -> A bsl B.

bitwise_shift_right(A, B) -> A bsr B.

from_base2(Str) ->
    try
        Res = binary_to_integer(Str, 2),
        {ok, Res}
    catch
        error:badarg -> {error, nil}
    end.

from_base8(Str) ->
    try
        Res = binary_to_integer(Str, 8),
        {ok, Res}
    catch
        error:badarg -> {error, nil}
    end.

from_base16(Str) ->
    try
        Res = binary_to_integer(Str, 16),
        {ok, Res}
    catch
        error:badarg -> {error, nil}
    end.

to_base2(Int) -> integer_to_binary(Int, 2).

to_base8(Int) -> integer_to_binary(Int, 8).

to_base16(Int) -> integer_to_binary(Int, 16).

File: client/bigi/bigi_ffi.mjs
import { Result$Ok, Result$Error, BitArray$BitArray } from "./gleam.mjs";
import { Order$Lt, Order$Eq, Order$Gt } from "../gleam_stdlib/gleam/order.mjs";
import { Endianness$isBigEndian, Signedness$isSigned } from "./bigi.mjs";

export function from(int) {
  return BigInt(int);
}

export function from_string(string) {
  try {
    return Result$Ok(BigInt(string));
  } catch {
    return Result$Error(undefined);
  }
}

export function from_bytes(bit_array, endianness, signedness) {
  if (bit_array.bitSize % 8 !== 0) {
    return Result$Error(undefined);
  }

  let value = 0n;

  // Read bytes as an unsigned integer value
  if (Endianness$isBigEndian(endianness)) {
    for (let i = 0; i < bit_array.byteSize; i++) {
      value = value * 256n + BigInt(bit_array.byteAt(i));
    }
  } else {
    for (let i = bit_array.byteSize - 1; i >= 0; i--) {
      value = value * 256n + BigInt(bit_array.byteAt(i));
    }
  }

  if (Signedness$isSigned(signedness)) {
    const byteSize = BigInt(bit_array.byteSize);

    const highBit = 2n ** (byteSize * 8n - 1n);

    // If the high bit is set and this is a signed integer, reinterpret as
    // two's complement
    if (value >= highBit) {
      value -= highBit * 2n;
    }
  }

  return Result$Ok(value);
}

export function to(bigint) {
  if (bigint > Number.MAX_SAFE_INTEGER || bigint < Number.MIN_SAFE_INTEGER) {
    return Result$Error(undefined);
  } else {
    return Result$Ok(Number(bigint));
  }
}

export function to_string(bigint) {
  return bigint.toString();
}

export function to_bytes(bigint, endianness, signedness, byte_count) {
  const bit_count = BigInt(byte_count * 8);

  if (bit_count < 8n) {
    return Result$Error(undefined);
  }

  let range_min = 0n;
  let range_max = 0n;

  // Error if the value is out of range for the available bits
  if (Signedness$isSigned(signedness)) {
    range_min = -(2n ** (bit_count - 1n));
    range_max = -range_min - 1n;
  } else {
    range_max = 2n ** bit_count - 1n;
  }

  if (bigint < range_min || bigint > range_max) {
    return Result$Error(undefined);
  }

  // Convert negative number to two's complement representation
  if (bigint < 0) {
    bigint = (1n << bit_count) + bigint;
  }

  const byteArray = new Uint8Array(byte_count);

  if (Endianness$isBigEndian(endianness)) {
    for (let i = byteArray.length - 1; i >= 0; i--) {
      const byte = bigint % 256n;
      byteArray[i] = Number(byte);
      bigint = (bigint - byte) / 256n;
    }
  } else {
    for (let i = 0; i < byteArray.length; i++) {
      const byte = bigint % 256n;
      byteArray[i] = Number(byte);
      bigint = (bigint - byte) / 256n;
    }
  }

  return Result$Ok(BitArray$BitArray(byteArray));
}

export function zero() {
  return 0n;
}

export function one() {
  return 1n;
}

export function n_one() {
  return -1n;
}

export function ten() {
  return 10n;
}

export function compare(a, b) {
  if (a < b) {
    return Order$Lt();
  } else if (a > b) {
    return Order$Gt();
  } else {
    return Order$Eq();
  }
}

export function absolute(bigint) {
  if (bigint < 0) {
    return -bigint;
  } else {
    return bigint;
  }
}

export function negate(bigint) {
  return -bigint;
}

export function add(a, b) {
  return a + b;
}

export function subtract(a, b) {
  return a - b;
}

export function multiply(a, b) {
  return a * b;
}

export function divide(a, b) {
  if (b === 0n) {
    return 0n;
  }

  return a / b;
}

export function divide_no_zero(a, b) {
  if (b === 0n) {
    return Result$Error(undefined);
  }

  return Result$Ok(divide(a, b));
}

export function remainder(a, b) {
  if (b === 0n) {
    return 0n;
  }

  return a % b;
}

export function remainder_no_zero(a, b) {
  if (b === 0n) {
    return Result$Error(undefined);
  }

  return Result$Ok(remainder(a, b));
}

export function modulo(a, b) {
  if (b === 0n) {
    return 0n;
  }

  return ((a % b) + b) % b;
}

export function modulo_no_zero(a, b) {
  if (b === 0n) {
    return Result$Error(undefined);
  }

  return Result$Ok(modulo(a, b));
}

export function power(a, b) {
  if (b < 0) {
    return Result$Error(undefined);
  }

  return Result$Ok(a ** b);
}

export function decode(dyn) {
  if (typeof dyn === "bigint") {
    return Result$Ok(dyn);
  } else {
    return Result$Error(0n);
  }
}

export function bitwise_and(a, b) {
  return a & b;
}

export function bitwise_exclusive_or(a, b) {
  return a ^ b;
}

export function bitwise_not(a) {
  return ~a;
}

export function bitwise_or(a, b) {
  return a | b;
}

export function bitwise_shift_left(a, b) {
  return a << BigInt(b);
}

export function bitwise_shift_right(a, b) {
  return a >> BigInt(b);
}

export function from_base2(str) {
  try {
    return Result$Ok(BigInt("0b" + str));
  } catch {
    return Result$Error(undefined);
  }
}

export function from_base8(str) {
  try {
    return Result$Ok(BigInt("0o" + str));
  } catch {
    return Result$Error(undefined);
  }
}

export function from_base16(str) {
  try {
    return Result$Ok(BigInt("0x" + str));
  } catch {
    return Result$Error(undefined);
  }
}

export function to_base2(int) {
  return int.toString(2);
}

export function to_base8(int) {
  return int.toString(8);
}

export function to_base16(int) {
  return int.toString(16);
}
